Need to re-analyze method groups

In `Binder.CreateConversion`, the bound method group node gets fixed to have a type (with oblivious annotations). During nullable analysis, we need to undo and re-do that with proper nullable annotation.

Note there is a parallel with lambda conversions. In `NullableWalker.GetUnboundLambda()` and `NullableWalker.ApplyConversion` we do this process for lambdas. I suspect we need something similar for method groups.

We should test all scenarios involving `NullableWalker.RemoveConversion()`. I suspect this logic of undoing effects of `Binder.CreateConversion` should be factored there. For example, `new[] { new Func<string?>(...), () => "" };`.

```C#
        [Fact]
        public void SuppressNullableWarning_LambdaInOverloadResolution()
        {
            var source =
@"class C
{
    static void Main(string? x)
    {
        var s = M(() => { return x; });
        s /*T:string?*/ .ToString(); // 1

        var s2 = M(() => { return x; }!); // suppressed
        s2 /*T:string?*/ .ToString(); // 2

        var s3 = M(M2);
        s3 /*T:string*/ .ToString(); // 3

        var s4 = M(M2!); // suppressed
        s4 /*T:string*/ .ToString(); // 4
    }
    static T M<T>(System.Func<T> x) => throw null;
    static string? M2() => throw null;
}";
            var comp = CreateCompilation(source, options: WithNonNullTypesTrue());
            comp.VerifyTypes();

            // TODO2
            // Missing warnings on s3 and s4
            comp.VerifyDiagnostics(
                // (6,9): warning CS8602: Possible dereference of a null reference.
                //         s /*T:string?*/ .ToString(); // 1
                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "s").WithLocation(6, 9),
                // (9,9): warning CS8602: Possible dereference of a null reference.
                //         s2 /*T:string?*/ .ToString(); // 2
                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "s2").WithLocation(9, 9)
                );
            CompileAndVerify(comp);
        }

```

--- a/src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/MethodTypeInference.cs
+++ b/src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/MethodTypeInference.cs
@@ -68,6 +68,28 @@ namespace Microsoft.CodeAnalysis.CSharp
 
     internal sealed class MethodTypeInferrer
     {
+        internal abstract class Extensions
+        {
+            internal static readonly Extensions Default = new DefaultExtensions();
+
+            internal abstract TypeWithAnnotations GetTypeWithAnnotations(BoundExpression expr);
+
+            internal abstract TypeWithAnnotations GetMethodGroupResultType(BoundMethodGroup group, MethodSymbol method);
+
+            private sealed class DefaultExtensions : Extensions
+            {
+                internal override TypeWithAnnotations GetTypeWithAnnotations(BoundExpression expr)
+                {
+                    return TypeWithAnnotations.Create(expr.Type);
+                }
+
+                internal override TypeWithAnnotations GetMethodGroupResultType(BoundMethodGroup group, MethodSymbol method)
+                {
+                    return method.ReturnTypeWithAnnotations;
+                }
+            }
+        }
+
         private enum InferenceResult
         {
             InferenceFailed,
@@ -91,7 +113,7 @@ namespace Microsoft.CodeAnalysis.CSharp
         private readonly ImmutableArray<TypeWithAnnotations> _formalParameterTypes;
         private readonly ImmutableArray<RefKind> _formalParameterRefKinds;
         private readonly ImmutableArray<BoundExpression> _arguments;
-        private readonly Func<BoundExpression, TypeWithAnnotations> _getTypeWithAnnotationOpt;
+        private readonly Extensions _extensions;
 
         private readonly TypeWithAnnotations[] _fixedResults;
         private readonly HashSet<TypeWithAnnotations>[] _exactBounds;
@@ -215,7 +237,7 @@ namespace Microsoft.CodeAnalysis.CSharp
             ImmutableArray<BoundExpression> arguments,// Required; in scenarios like method group conversions where there are
                                                       // no arguments per se we cons up some fake arguments.
             ref HashSet<DiagnosticInfo> useSiteDiagnostics,
-            Func<BoundExpression, TypeWithAnnotations> getTypeWithAnnotationOpt = null)
+            Extensions extensions = null)
         {
             Debug.Assert(!methodTypeParameters.IsDefault);
             Debug.Assert(methodTypeParameters.Length > 0);
@@ -240,7 +262,7 @@ namespace Microsoft.CodeAnalysis.CSharp
                 formalParameterTypes,
                 formalParameterRefKinds,
                 arguments,
-                getTypeWithAnnotationOpt);
+                extensions);
             return inferrer.InferTypeArgs(binder, ref useSiteDiagnostics);
         }
 
@@ -260,7 +282,7 @@ namespace Microsoft.CodeAnalysis.CSharp
             ImmutableArray<TypeWithAnnotations> formalParameterTypes,
             ImmutableArray<RefKind> formalParameterRefKinds,
             ImmutableArray<BoundExpression> arguments,
-            Func<BoundExpression, TypeWithAnnotations> getTypeWithAnnotationOpt)
+            Extensions extensions)
         {
             _conversions = conversions;
             _methodTypeParameters = methodTypeParameters;
@@ -268,7 +290,7 @@ namespace Microsoft.CodeAnalysis.CSharp
             _formalParameterTypes = formalParameterTypes;
             _formalParameterRefKinds = formalParameterRefKinds;
             _arguments = arguments;
-            _getTypeWithAnnotationOpt = getTypeWithAnnotationOpt;
+            _extensions = extensions ?? Extensions.Default;
             _fixedResults = new TypeWithAnnotations[methodTypeParameters.Length];
             _exactBounds = new HashSet<TypeWithAnnotations>[methodTypeParameters.Length];
             _upperBounds = new HashSet<TypeWithAnnotations>[methodTypeParameters.Length];
@@ -549,7 +571,7 @@ namespace Microsoft.CodeAnalysis.CSharp
                 // Either the argument is not a tuple literal, or we were unable to do the inference from its elements, let's try to infer from argument type
                 if (IsReallyAType(argument.Type))
                 {
-                    ExactOrBoundsInference(kind, GetTypeWithAnnotations(argument), target, ref useSiteDiagnostics);
+                    ExactOrBoundsInference(kind, _extensions.GetTypeWithAnnotations(argument), target, ref useSiteDiagnostics);
                 }
             }
         }
@@ -1190,7 +1212,7 @@ namespace Microsoft.CodeAnalysis.CSharp
             }
             // SPEC: * Otherwise, if E is an expression with type U then a lower-bound
             // SPEC:   inference is made from U to T.
-            var sourceType = GetTypeWithAnnotations(expression);
+            var sourceType = _extensions.GetTypeWithAnnotations(expression);
             if (sourceType.HasType)
             {
                 LowerBoundInference(sourceType, target, ref useSiteDiagnostics);
@@ -1274,20 +1296,16 @@ namespace Microsoft.CodeAnalysis.CSharp
             }
 
             var returnType = MethodGroupReturnType(binder, (BoundMethodGroup)source, fixedDelegateParameters, delegateInvokeMethod.RefKind, ref useSiteDiagnostics);
-            if ((object)returnType == null || returnType.IsVoidType())
+            if (returnType.IsDefault || returnType.IsVoidType())
             {
                 return false;
             }
 
-            // https://github.com/dotnet/roslyn/issues/33635 : We should preserve the return nullability from the
-            // selected method of the method group, possibly turning oblivious into non-null.
-            NullableAnnotation returnIsNullable = NullableAnnotation.Oblivious;
-            LowerBoundInference(TypeWithAnnotations.Create(returnType, returnIsNullable), delegateReturnType, ref useSiteDiagnostics);
-
+            LowerBoundInference(returnType, delegateReturnType, ref useSiteDiagnostics);
             return true;
         }
 
-        private static TypeSymbol MethodGroupReturnType(
+        private TypeWithAnnotations MethodGroupReturnType(
             Binder binder, BoundMethodGroup source,
             ImmutableArray<ParameterSymbol> delegateParameters,
             RefKind delegateRefKind,
@@ -1301,7 +1319,7 @@ namespace Microsoft.CodeAnalysis.CSharp
                 // Since we are trying to infer the return type, it is not an input to resolving the method group
                 returnType: null);
 
-            TypeSymbol type = null;
+            TypeWithAnnotations type = default;
 
             // The resolution could be empty (e.g. if there are no methods in the BoundMethodGroup).
             if (!resolution.IsEmpty)
@@ -1309,7 +1327,7 @@ namespace Microsoft.CodeAnalysis.CSharp
                 var result = resolution.OverloadResolutionResult;
                 if (result.Succeeded)
                 {
-                    type = result.BestResult.Member.ReturnType;
+                    type = _extensions.GetMethodGroupResultType(source, result.BestResult.Member);
                 }
             }
 
@@ -2437,15 +2455,6 @@ OuterBreak:
             return best;
         }
 
-        private TypeWithAnnotations GetTypeWithAnnotations(BoundExpression expr)
-        {
-            if (_conversions.IncludeNullability && _getTypeWithAnnotationOpt != null)
-            {
-                return _getTypeWithAnnotationOpt(expr);
-            }
-            return TypeWithAnnotations.Create(expr.Type);
-        }
-
         internal static TypeWithAnnotations Merge(TypeWithAnnotations first, TypeWithAnnotations second, VarianceKind variance, ConversionsBase conversions)
         {
             var merged = MergeTupleNames(MergeDynamic(first, second, conversions.CorLibrary), second);
@@ -2706,7 +2715,7 @@ OuterBreak:
                 constructedFromMethod.GetParameterTypes(),
                 constructedFromMethod.ParameterRefKinds,
                 arguments,
-                getTypeWithAnnotationOpt: null);
+                extensions: null);
 
             if (!inferrer.InferTypeArgumentsFromFirstArgument(ref useSiteDiagnostics))
             {
@@ -2732,7 +2741,7 @@ OuterBreak:
             {
                 return false;
             }
-            LowerBoundInference(GetTypeWithAnnotations(argument), dest, ref useSiteDiagnostics);
+            LowerBoundInference(_extensions.GetTypeWithAnnotations(argument), dest, ref useSiteDiagnostics);
             // Now check to see that every type parameter used by the first
             // formal parameter type was successfully inferred.
             for (int iParam = 0; iParam < _methodTypeParameters.Length; ++iParam)

--- a/src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs
+++ b/src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs
@@ -3467,7 +3467,7 @@ namespace Microsoft.CodeAnalysis.CSharp
                 parameterRefKinds,
                 arguments,
                 ref useSiteDiagnostics,
-                getTypeWithAnnotationOpt: s_getTypeWithAnnotations);
+                new MethodInferenceExtensions(this));
 
             if (!result.Success)
             {
@@ -3477,8 +3477,32 @@ namespace Microsoft.CodeAnalysis.CSharp
             return definition.Construct(result.InferredTypeArguments);
         }
 
-        private readonly static Func<BoundExpression, TypeWithAnnotations> s_getTypeWithAnnotations =
-            (expr) => TypeWithAnnotations.Create(expr.Type, GetNullableAnnotation(expr));
+        private sealed class MethodInferenceExtensions : MethodTypeInferrer.Extensions
+        {
+            private readonly NullableWalker _walker;
+
+            internal MethodInferenceExtensions(NullableWalker walker)
+            {
+                _walker = walker;
+            }
+
+            internal override TypeWithAnnotations GetTypeWithAnnotations(BoundExpression expr)
+            {
+                return TypeWithAnnotations.Create(expr.Type, GetNullableAnnotation(expr));
+            }
+
+            internal override TypeWithAnnotations GetMethodGroupResultType(BoundMethodGroup group, MethodSymbol method)
+            {
+                if (_walker.TryGetMethodGroupReceiverNullability(group.ReceiverOpt, out TypeWithState receiverType))
+                {
+                    if (!method.IsStatic)
+                    {
+                        method = (MethodSymbol)AsMemberOfType(receiverType.Type, method);
+                    }
+                }
+                return method.ReturnTypeWithAnnotations;
+            }
+        }
 
         private ImmutableArray<BoundExpression> GetArgumentsForMethodTypeInference(ImmutableArray<BoundExpression> arguments, ImmutableArray<VisitArgumentResult> argumentResults)
         {

--- a/src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs
+++ b/src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs
@@ -2133,10 +2133,10 @@ class Test
         s2 /*T:string?*/ .ToString(); // 2
 
         var s3 = M(M2);
-        s3 /*T:string!*/ .ToString(); // 3
+        s3 /*T:string?*/ .ToString(); // 3
 
         var s4 = M(M2!); // suppressed
-        s4 /*T:string!*/ .ToString(); // 4
+        s4 /*T:string?*/ .ToString(); // 4
     }
     static T M<T>(System.Func<T> x) => throw null!;
     static string? M2() => throw null!;
@@ -2144,17 +2144,19 @@ class Test
             var comp = CreateCompilation(source, options: WithNonNullTypesTrue());
             comp.VerifyTypes();
 
-            // Missing warnings on s3 and s4
-            // Tracked by https://github.com/dotnet/roslyn/issues/32697
-
             comp.VerifyDiagnostics(
                 // (6,9): warning CS8602: Dereference of a possibly null reference.
                 //         s /*T:string?*/ .ToString(); // 1
                 Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "s").WithLocation(6, 9),
                 // (9,9): warning CS8602: Dereference of a possibly null reference.
                 //         s2 /*T:string?*/ .ToString(); // 2
-                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "s2").WithLocation(9, 9)
-                );
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "s2").WithLocation(9, 9),
+                // (12,9): warning CS8602: Dereference of a possibly null reference.
+                //         s3 /*T:string?*/ .ToString(); // 3
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "s3").WithLocation(12, 9),
+                // (15,9): warning CS8602: Dereference of a possibly null reference.
+                //         s4 /*T:string?*/ .ToString(); // 4
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "s4").WithLocation(15, 9));
             CompileAndVerify(comp);
         }
 
@@ -2268,12 +2270,12 @@ class A
     void M()
     {
         IEnumerable<string?> c = null!;
-        c.S(G)/*T:IEnumerable<object!>!*/; // 1
-        c.S(G!)/*T:IEnumerable<object!>!*/;
+        c.S(G)/*T:System.Collections.Generic.IEnumerable<object!>!*/; // 1
+        c.S(G!)/*T:System.Collections.Generic.IEnumerable<object!>!*/;
 
         IEnumerable<string> c2 = null!;
-        c2.S(G)/*T:IEnumerable<object!>!*/;
-        c2.S(G2)/*T:IEnumerable<object!>!*/;
+        c2.S(G)/*T:System.Collections.Generic.IEnumerable<object!>!*/;
+        c2.S(G2)/*T:System.Collections.Generic.IEnumerable<object!>!*/;
     }
     static object G(string s) => throw null!;
     static object G2(string? s) => throw null!;
@@ -2292,8 +2294,7 @@ static class E
                 //         c.S(G)/*T:IEnumerable<object!>!*/; // 1
                 Diagnostic(ErrorCode.WRN_NullabilityMismatchInParameterTypeOfTargetDelegate, "G").WithArguments("s", "object A.G(string s)", "System.Func<string?, object>").WithLocation(8, 13)
                 );
-            // https://github.com/dotnet/roslyn/issues/33635
-            // comp.VerifyTypes();
+            comp.VerifyTypes();
         }
 
         [Fact, WorkItem(31370, "https://github.com/dotnet/roslyn/issues/31370")]
@@ -73004,8 +73005,7 @@ public delegate (T, U) MyDelegate<T, U>(U u);
 ";
             var comp = CreateCompilation(new[] { source }, options: WithNonNullTypesTrue());
             comp.VerifyDiagnostics();
-            // https://github.com/dotnet/roslyn/issues/33637
-            //comp.VerifyTypes();
+            comp.VerifyTypes();
         }
 
         [WorkItem(33638, "https://github.com/dotnet/roslyn/issues/33638")]

